#### Викулин Иван Сергевич РТ5-61Б
### Вариант 4

### Задание
Для заданного набора данных постройте модели классификации или регрессии. Для построения моделей используйте методы Дерево решений и Градиентный бустинг. Оцените качество моделей на основе подходящих метрик качества (не менее двух метрик). Какие метрики качества Вы использовали и почему? Какие выводы Вы можете сделать о качестве построенных моделей? Для построения моделей необходимо выполнить требуемую предобработку данных: заполнение пропусков, кодирование категориальных признаков, и т.д.

### Полученный датасет Toy Dataset
Вымышленный набор данных для исследовательского анализа данных (EDA) и для тестирования простых моделей прогнозирования.

Этот учебный набор данных содержит 150000 строк и 6 столбцов.

Столбцы

Номер: Простой индексный номер для каждой строки

Город: Местоположение человека (Даллас, Нью-Йорк, Лос-Анджелес, Маунтин-Вью, Бостон, Вашингтон, Сан-Диего и Остин)

Пол: Пол человека (Мужчина или Женщина)

Возраст: Возраст человека (в диапазоне от 25 до 65 лет)

Доход: Годовой доход человека (в диапазоне от -674 до 177175)

Болезнь: Больной ли человек? (Да или Нет)

Примечание: Все данные вымышлены. Данные сгенерированы таким образом, чтобы их распределения были удобны для статистического анализа.

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
sns.set(style="ticks")

In [4]:
data = pd.read_csv('toy_dataset.csv')

In [5]:
data.dtypes

Number       int64
City        object
Gender      object
Age          int64
Income     float64
Illness     object
dtype: object

In [7]:
data.shape

(150000, 6)

In [8]:
data.isnull().sum()

Number     0
City       0
Gender     0
Age        0
Income     0
Illness    0
dtype: int64

Пропусков нет

In [9]:
data.head()

Number    City Gender  Age   Income Illness
0       1  Dallas   Male   41  40367.0      No
1       2  Dallas   Male   54  45084.0      No
2       3  Dallas   Male   42  52483.0      No
3       4  Dallas   Male   40  40941.0      No
4       5  Dallas   Male   46  50289.0      No

### Кодироание категориальных признаков

In [19]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder

In [11]:
data.City.unique()

array(['Dallas', 'New York City', 'Los Angeles', 'Mountain View',
       'Boston', 'Washington D.C.', 'San Diego', 'Austin'], dtype=object)

In [29]:
df = pd.DataFrame(data)

In [30]:
ohe = OneHotEncoder(sparse_output=False)
city_encoded = ohe.fit_transform(df[['City']])
city_encoded_df = pd.DataFrame(city_encoded, columns=ohe.get_feature_names_out(['City']))

In [31]:
df = pd.concat([df, city_encoded_df], axis=1).drop('City', axis=1)
print(df)

        Number  Gender  Age    Income Illness  City_Austin  City_Boston  \
0            1    Male   41   40367.0      No          0.0          0.0   
1            2    Male   54   45084.0      No          0.0          0.0   
2            3    Male   42   52483.0      No          0.0          0.0   
3            4    Male   40   40941.0      No          0.0          0.0   
4            5    Male   46   50289.0      No          0.0          0.0   
...        ...     ...  ...       ...     ...          ...          ...   
149995  149996    Male   48   93669.0      No          1.0          0.0   
149996  149997    Male   25   96748.0      No          1.0          0.0   
149997  149998    Male   26  111885.0      No          1.0          0.0   
149998  149999    Male   25  111878.0      No          1.0          0.0   
149999  150000  Female   37   87251.0      No          1.0          0.0   

        City_Dallas  City_Los Angeles  City_Mountain View  City_New York City  \
0               1.

In [17]:
df.City.unique()

array([1, 0])

In [33]:
le_gender = LabelEncoder()
df['Gender'] = le_gender.fit_transform(df['Gender'])

In [34]:
le_illness = LabelEncoder()
df['Illness'] = le_illness.fit_transform(df['Illness'])

In [35]:
print(df)

        Number  Gender  Age    Income  Illness  City_Austin  City_Boston  \
0            1       1   41   40367.0        0          0.0          0.0   
1            2       1   54   45084.0        0          0.0          0.0   
2            3       1   42   52483.0        0          0.0          0.0   
3            4       1   40   40941.0        0          0.0          0.0   
4            5       1   46   50289.0        0          0.0          0.0   
...        ...     ...  ...       ...      ...          ...          ...   
149995  149996       1   48   93669.0        0          1.0          0.0   
149996  149997       1   25   96748.0        0          1.0          0.0   
149997  149998       1   26  111885.0        0          1.0          0.0   
149998  149999       1   25  111878.0        0          1.0          0.0   
149999  150000       0   37   87251.0        0          1.0          0.0   

        City_Dallas  City_Los Angeles  City_Mountain View  City_New York City  \
0     

### Разделение на обучающую и тестовую выборки

Будем решать задачу классификации, с целевым признаком - наличие болезни

In [36]:
from sklearn.model_selection import train_test_split

In [37]:
X = df.drop("Illness", axis=1)
y = df["Illness"]

In [38]:
print(X.head(), "\n")
print(y.head())

   Number  Gender  Age   Income  City_Austin  City_Boston  City_Dallas  \
0       1       1   41  40367.0          0.0          0.0          1.0   
1       2       1   54  45084.0          0.0          0.0          1.0   
2       3       1   42  52483.0          0.0          0.0          1.0   
3       4       1   40  40941.0          0.0          0.0          1.0   
4       5       1   46  50289.0          0.0          0.0          1.0   

   City_Los Angeles  City_Mountain View  City_New York City  City_San Diego  \
0               0.0                 0.0                 0.0             0.0   
1               0.0                 0.0                 0.0             0.0   
2               0.0                 0.0                 0.0             0.0   
3               0.0                 0.0                 0.0             0.0   
4               0.0                 0.0                 0.0             0.0   

   City_Washington D.C.  
0                   0.0  
1                   0.0  
2 

In [39]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=9)

In [40]:
X_train.shape, y_train.shape,

((112500, 12), (112500,))

In [41]:
X_test.shape, y_test.shape

((37500, 12), (37500,))

### Дерево решений

In [65]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, recall_score, precision_score

In [43]:
model = DecisionTreeClassifier(random_state=3)
model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=3)

In [44]:
y_pred = model.predict(X_test)

In [66]:
accuracy = accuracy_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)

In [67]:
print(f'Точность: {accuracy}')
print(f'precision: {precision}')
print(f'Полнота: {recall}')

Точность: 0.8443733333333333
precision: 0.09066901408450705
Полнота: 0.10144451739986868


### Градиентный бустинг

In [54]:
from sklearn.ensemble import  GradientBoostingClassifier

In [55]:
gradient_boosting_classifier= GradientBoostingClassifier()
gradient_boosting_classifier.fit(X_train, y_train)

GradientBoostingClassifier()

In [56]:
y_grad_pred = gradient_boosting_classifier.predict(X_test)

In [69]:
grad_accuracy = accuracy_score(y_test, y_grad_pred)
grad_recall = recall_score(y_test, y_grad_pred)
grad_precision = precision_score(y_test, y_grad_pred)

In [70]:
print(f'Точность: {grad_accuracy}')
print(f'precision: {grad_precision}')
print(f'Полнота: {grad_recall}')


Точность: 0.9187733333333333
precision: 0.5
Полнота: 0.0006565988181221273


## Вывод


Градиентный бустинг имеет значительно более высокую точность, чем дерево решений (91% и 84% соответственно), в случае с precision - 50% и 9.1%.
Однако полнота градиентного бустинга очень низкая, менее 0.1%, а это означает что он очень плохо обнаруживает положительные случаи.
Дерево решений имеет более низкую точность, но зато более высокую полноту, что может быть полезным в ситуациях, где важно обнаружить все положительные случаи, даже за счет увеличения числа ложноположительных срабатываний.